lib/spack/spack/package.py

```diff
@@ -1217,6 +1236,15 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class PythonExtension(Package):
+    def setup_dependent_environment(self, env, dependent_spec):
+        pass
+
+    def setup_environment(self, env):
+        site_packages = glob.glob(join_path(self.spec.prefix.lib, "python*/site-packages"))
+        if site_packages:
+            env.prepend_path('PYTHONPATH', site_packages[0])
```

Comment:
@tgamblin I think I may be missing some bits... For `PYTHONPATH` I am trying to reproduce the behavior of develop, but moving the extension specific knowledge out of `modules.py`. The code that in develop takes care of adding `PYTHONPATH` in `modules.py` is:
```python
if self.spec.name != 'python':
    site_packages = glob(join_path(self.spec.prefix.lib, "python*/site-packages"))
    if site_packages:
        add_path('PYTHONPATH', site_packages[0])
```
This logic is triggered only for modules (be it dk or tcl), because as you said, when things are installed it's the `setup_dependent_environment` of python that kicks in:
```python
# Set PYTHONPATH to include site-packages dir for the
# extension and any other python extensions it depends on.
python_paths = []
for d in ext_spec.traverse():
    if d.package.extends(self.spec):
        python_paths.append(os.path.join(d.prefix, self.site_packages_dir))
    os.environ['PYTHONPATH'] = ':'.join(python_paths)
```
The class above should be a way to reproduce the same behavior and transfer the responsibility to the packages that implements python extensions. At install time it's the `setup_dependent_environment` of python that is executed, and the implementation it's the same as the one in develop... so I believe the code here should behave as correctly as in develop. Do you see a fault in the reasoning?

As far as convenience is concerned I agree with you that this is not the optimal solution. I guess anyhow that for most python packages the implementation of `setup_environment` will be the same and given the number of packages already in spack we should definitely find a way of factoring the code in a unique place. The only idea that comes to my mind right now would be patching the default implementation of `setup_environment` and `setup_dependent_environment` on the fly when an `extends` directive is used. This is also kind of nasty, and if you have a better idea I will be glad to implement it :smile: